pkg/compactor/compactor.go

```diff
@@ -383,42 +392,39 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
 	c.bucketClient = bucketindex.BucketWithGlobalMarkers(c.bucketClient)
 
 	// Initialize the compactors ring if sharding is enabled.
-	lifecyclerCfg := c.compactorCfg.ShardingRing.ToLifecyclerConfig()
-	c.ringLifecycler, err = ring.NewLifecycler(lifecyclerCfg, ring.NewNoopFlushTransferer(), "compactor", CompactorRingKey, false, c.logger, prometheus.WrapRegistererWithPrefix("cortex_", c.registerer))
+	c.ring, c.ringLifecycler, err = newRingAndLifecycler(c.compactorCfg.ShardingRing, c.logger, c.registerer)
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring lifecycler")
+		return err
 	}
 
-	c.ring, err = ring.New(lifecyclerCfg.RingConfig, "compactor", CompactorRingKey, c.logger, prometheus.WrapRegistererWithPrefix("cortex_", c.registerer))
+	c.ringSubservices, err = services.NewManager(c.ringLifecycler, c.ring)
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring")
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	c.ringSubservices, err = services.NewManager(c.ringLifecycler, c.ring)
-	if err == nil {
-		c.ringSubservicesWatcher = services.NewFailureWatcher()
-		c.ringSubservicesWatcher.WatchManager(c.ringSubservices)
-
-		err = services.StartManagerAndAwaitHealthy(ctx, c.ringSubservices)
+	c.ringSubservicesWatcher = services.NewFailureWatcher()
+	c.ringSubservicesWatcher.WatchManager(c.ringSubservices)
+	if err = c.ringSubservices.StartAsync(ctx); err != nil {
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	if err != nil {
+	ctxTimeout, cancel = context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
+	defer cancel()
+	if err = c.ringSubservices.AwaitHealthy(ctxTimeout); err != nil {
 		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	// If sharding is enabled we should wait until this instance is
-	// ACTIVE within the ring. This MUST be done before starting the
-	// any other component depending on the users scanner, because the
-	// users scanner depends on the ring (to check whether an user belongs
-	// to this shard or not).
+	// If sharding is enabled we should wait until this instance is ACTIVE within the ring. This
+	// MUST be done before starting any other component depending on the users scanner, because
+	// the users scanner depends on the ring (to check whether a user belongs to this shard or not).
 	level.Info(c.logger).Log("msg", "waiting until compactor is ACTIVE in the ring")
 
-	ctxWithTimeout, cancel := context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
+	ctxTimeout, cancel = context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
```

Comment:
I don't think we should create a new `ctxTimeout` here, because we apply `WaitActiveInstanceTimeout` two times, while I think we should just reuse the previous `ctxTimeout`.

Reply:
The way `WithTimeout` is implemented, any time spent waiting for the first place we use it takes time away from the second place we use it. If I implement your suggestion, starting and waiting for the instance to become active will only have N seconds total to complete. The way it's currently written, starting has N seconds to complete and waiting for the instance to become active has N seconds to complete. I can make that change but want to confirm that's what you intend.

Reply:
> I can make that change but want to confirm that's what you intend.

That's what I meant (all in all it mimics the previous behaviour) but I honestly don't feel very strong about it. The whole purpose of this timeout was just to avoid to waiting indefinitely. Keep your code as is 👍